### ISSUE TITLE
`tuple_array_conversion` false positive on destructured tuple values

### ISSUE THREAD

**Issue by MingweiSamuel** [rust-lang/rust-clippy#11144]
### Summary

The `tuple_array_conversion` seems to be way to zealous in detecting the original tuple identifiers, generating false positives even when the names are assigned through destructuring. These all trigger it even though it makes no sense to:
```rust
let (a, (_b, c)) = get_tuple_a();
let _ = [a, c];

let (a, [_b, c]) = get_tuple_b();
let _ = [a, c];

let (a, Wrap(b)) = get_tuple_c();
let _ = [a, b];
```

full code below

### Lint Name

tuple_array_conversion

### Reproducer

full code:
```rust
fn get_tuple_a() -> (usize, (usize, usize)) {
    (1, (2, 3))
}

fn get_tuple_b() -> (usize, [usize; 2]) {
    (1, [2, 3])
}

struct Wrap<T>(pub T);
fn get_tuple_c() -> (usize, Wrap<usize>) {
    (1, Wrap(2))
}

fn main() {
    let (a, (_b, c)) = get_tuple_a();
    let _ = [a, c];

    let (a, [_b, c]) = get_tuple_b();
    let _ = [a, c];
    
    let (a, Wrap(b)) = get_tuple_c();
    let _ = [a, b];
}
```

```rust
    Checking playground v0.0.1 (/playground)
warning: it looks like you're trying to convert a tuple to an array
  --> src/main.rs:16:13
   |
16 |     let _ = [a, c];
   |             ^^^^^^
   |
   = help: use `.into()` instead, or `<[T; N]>::from` if type annotations are needed
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#tuple_array_conversions
   = note: `#[warn(clippy::tuple_array_conversions)]` on by default

warning: it looks like you're trying to convert a tuple to an array
  --> src/main.rs:19:13
   |
19 |     let _ = [a, c];
   |             ^^^^^^
   |
   = help: use `.into()` instead, or `<[T; N]>::from` if type annotations are needed
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#tuple_array_conversions

warning: it looks like you're trying to convert a tuple to an array
  --> src/main.rs:22:13
   |
22 |     let _ = [a, b];
   |             ^^^^^^
   |
   = help: use `.into()` instead, or `<[T; N]>::from` if type annotations are needed
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#tuple_array_conversions

warning: `playground` (bin "playground") generated 3 warnings
    Finished dev [unoptimized + debuginfo] target(s) in 0.44s
```

playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=504c4f33d4880029600437ce2a068acf

### Version

```text
1.73.0-nightly
(2023-07-11 993deaa0bf8bab9dd3ea)
```

**Comment by Centri3** [comment]
This lint should probably be rewritten. Currently it checks if all items in the resulting array/tuple come from the same pattern, which is a tuple (which these do btw), rather than just the pattern's first parent.

**Comment by GoldsteinE** [comment]
A case I encountered:

```rust
use std::path::PathBuf;

fn main() {
    let v = std::env::var("TEST");
    let (a, b) = match &v {
        Ok(x) => ("a", &**x),
        Err(_) => ("b", "c"),
    };
    let _x = PathBuf::from_iter([a, b]);
    println!("{a}");
}
```

I don’t think replacing `(a, b)` with a single array makes sense in this case, since I use `a` afterwards.

**Comment by Centri3** [comment]
This is already mentioned in #11124
